skyportal/models.py

```diff
@@ -1758,6 +1759,231 @@ def api_class(self):
+
+    validity_window_start = sa.Column(
+        sa.DateTime,
+        nullable=False,
+        default=lambda: datetime.now(),
+        doc='Start of validity window',
+    )
+
+    validity_window_end = sa.Column(
+        sa.DateTime,
+        nullable=False,
+        default=lambda: datetime.now() + timedelta(1),
+        doc='End of validity window',
+    )
+
+    plan_args = sa.Column(sa.JSON, nullable=False, doc='Plan arguments')
+
+    class Status(enum.IntEnum):
+        WORKING = 0
+        READY = 1
+        SUBMITTED = 2
+
+    status = sa.Column(
+        sa.Enum(Status), default=Status.WORKING, nullable=False, doc='Plan status'
+    )
+
+    planned_observations = relationship("PlannedObservation")
+
+    @property
+    def start_observation(self):
+        """Time of the first planned observation."""
+        if self.planned_observations:
+            return self.planned_observations[0].obstime
+        else:
+            return None
+
+    @hybrid_property
+    def num_observations(self):
+        """Number of planned observation."""
+        return len(self.planned_observations)
+
+    @num_observations.expression
+    def num_observations(cls):
+        """Number of planned observation."""
+        return cls.planned_observations.count()
+
+    @property
+    def num_observations_per_filter(self):
+        """Number of planned observation per filter."""
+        filters = list(Telescope.query.get(self.telescope).filters)
+        nepochs = np.zeros(
+            len(filters),
+        )
+        bands = {1: 'g', 2: 'r', 3: 'i', 4: 'z', 5: 'J'}
+        for planned_observation in self.planned_observations:
+            filt = bands[planned_observation.filter_id]
+            idx = filters.index(filt)
+            nepochs[idx] = nepochs[idx] + 1
+        nobs_per_filter = []
+        for ii, filt in enumerate(filters):
+            nobs_per_filter.append("%s: %d" % (filt, nepochs[ii]))
+        return " ".join(nobs_per_filter)
+
+    @property
+    def total_time(self):
+        """Total observation time (seconds)."""
+        return sum(_.exposure_time for _ in self.planned_observations)
+
+    @property
+    def tot_time_with_overheads(self):
+        overhead = sum(_.overhead_per_exposure for _ in self.planned_observations)
+        return overhead + self.total_time
+
+    @property
+    def ipix(self):
+        return {
+            i
+            for planned_observation in self.planned_observations
+            if planned_observation.field.ipix is not None
+            for i in planned_observation.field.ipix
+        }
+
+    @property
+    def area(self):
+        nside = Localization.nside
+        return hp.nside2pixarea(nside, degrees=True) * len(self.ipix)
+
+    def get_probability(self, localization):
+        ipix = np.asarray(list(self.ipix))
+        if len(ipix) > 0:
+            return localization.flat_2d[ipix].sum()
+        else:
+            return 0.0
+
+
+class PlannedObservation(Base):
```

Comment:
Should these be named more explicitly to distinguish them from the existing `ObservingRun`, etc, classes?

Reply:
Suggestion? They do have Planned in the title.

Reply:
Well these are quite distinct from followup requests, right? Something to disambiguate would be nice

Reply:
ObservingPlan -> EventObservingPlan ?